[pytorch] fix THAllocator.cpp

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19762 [pytorch] CMakeLists changes to enable libtorch for Android
* #19942 [pytorch] rename BUILD_ATEN_MOBILE to INTERN_BUILD_MOBILE and make it private
* #19761 [pytorch] add new macro TORCH_MOBILE for libtorch mobile build
* #19760 [pytorch] fix labs warning in THTensorMoreMath.cpp
* **#19759 [pytorch] fix THAllocator.cpp**
* #19758 [pytorch] remove C10_MOBILE from LegacyTypeDispatch.cpp

Summary:
Update THRefcountedMapAllocator constructor/close() to fix build error
when _WIN32 and HAVE_MMAP are undefined - which case doesn't seem to be
covered by any CI.

Discovered by Thomas in https://github.com/pytorch/pytorch/pull/16242
Seems to be a good fix regardless of android build.

Test Plan:
Build with stacked diffs

Differential Revision: [D15087651](https://our.internmc.facebook.com/intern/diff/D15087651)

ljk53 committed:
[pytorch] fix THAllocator.cpp

Summary:
Update THRefcountedMapAllocator constructor/close() to fix build error
when _WIN32 and HAVE_MMAP are undefined - which case doesn't seem to be
covered by any CI.

Discovered by Thomas in https://github.com/pytorch/pytorch/pull/16242
Seems to be a good fix regardless of android build.

Test Plan:
Build with stacked diffs

ljk53 committed:
Update on "[pytorch] fix THAllocator.cpp"

apaszke commented on aten/src/TH/THAllocator.cpp:
```diff
@@ -496,16 +496,23 @@ int THRefcountedMapAllocator::decref()
-THRefcountedMapAllocator::THRefcountedMapAllocator(WithFd, const char *filename, int fd, int flags, size_t size) {
+THRefcountedMapAllocator::THRefcountedMapAllocator(WithFd, const char *filename, int fd, int flags, size_t size)
+  : THRefcountedMapAllocatorArgCheck(flags),
+    THMapAllocator(WITH_FD, filename, flags, fd, size + TH_ALLOC_ALIGNMENT)
```
Those calls seem unnecessary given that we raise an error anyway?

ljk53 replied:
You are right that these calls should not be called - but seems we still need make these changes to make it compile. In some earlier diff we changed the base class which no longer takes default constructor - other call sites were updated accordingly but this one was missed and it's not covered by integration test.

apaszke replied:
Oh, I see. Makes sense!

ljk53 committed:
Update on "[pytorch] fix THAllocator.cpp"

[pytorch] fix THAllocator.cpp

Summary:
Update THRefcountedMapAllocator constructor/close() to fix build error
when _WIN32 and HAVE_MMAP are undefined - which case doesn't seem to be
covered by any CI.

Discovered by Thomas in https://github.com/pytorch/pytorch/pull/16242
Seems to be a good fix regardless of android build.

Test Plan:
Build with stacked diffs

gh-metadata: pytorch pytorch 19759 gh/ljk53/11/head
